File: leetcode/Python/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, node, maxVal):
        count=0
        if not node:
            return count
        if node.val>=maxVal:
            count+=1
            maxVal=node.val
        count+=self.dfs(node.left,maxVal)+self.dfs(node.right,maxVal)
        return count
            
    def goodNodes(self, root: TreeNode) -> int:
        return self.dfs(root, -10000)

File: leetcode/C++/count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node, int maxVal){
        int count=0;
        if(node==nullptr) return count;
        if(node->val>=maxVal) {
            ++count;
            maxVal=node->val;
        }
        count+=dfs(node->left,maxVal)+dfs(node->right,maxVal);
        return count;
    }
    int goodNodes(TreeNode* root) {
        return dfs(root,-100000);
    }
};